File: 116-populating-next-right-pointers-in-each-node/Case1_DFS.cpp
//Time: 19 ms (85.09%), Space: 16.6 MB (96.67%) - LeetHub

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    void dfs(Node* node) {
        if(!node->left) return;
        
        node->left->next = node->right;        
        if(node->next)
            node->right->next = node->next->left;
        
        dfs(node->right);
        dfs(node->left);
    }
    Node* connect(Node* root) {
        if(root)
            dfs(root);
        return root;
    }
};

File: 785-is-graph-bipartite/Case1_DFS.java
//Time: 1 ms (83.33%), Space: 53.7 MB (47.00%) - LeetHub

class Solution {
    int[] visited;
    
    private boolean DFS(int[][] graph, int x) {
        
        for (int next : graph[x]) {
            if (visited[next] == 0) {
                visited[next] = 3 - visited[x];
                if (!DFS(graph, next))
                    return false;
            }else if (visited[next] == visited[x])
                return false;
        }
        return true;
    }
    public boolean isBipartite(int[][] graph) {
    
        visited = new int[graph.length];
        
        for (int i = 0; i < graph.length; i++) 
            if (visited[i] == 0) {
                visited[i] = 1;
                if (!DFS(graph, i))
                    return false;
            }
        
        return true;
    }
}
